src/Microcosm.js

```diff
@@ -20,104 +19,50 @@ let Microcosm = function() {
+
+  // A place to keep all developer warnings
+  this.addPlugin(Debug)
+
+  // Manage key lifecycle hooks for state
+  this.addPlugin(State)
+
+  // Keep track of history as a tree
+  this.addPlugin(History, options)
 }
 
 Microcosm.prototype = {
   constructor: Microcosm,
 
   getInitialState() {
-    return dispatch(this.stores, {}, Transaction(lifecycle.willStart, this.state))
+    return this.lifecycle(lifecycle.willStart)
   },
 
-  /**
-   * Dispatch all open transactions upon base state to determine
-   * a new state. This is the state exposed to the outside world.
-   */
   rollforward() {
-    this.state = this.history.branch().reduce((state, transaction) => {
-      return dispatch(this.stores, state, transaction)
-    }, this.base)
+    this.state = this.lifecycle(lifecycle.willRollforward)
 
     this.emit(this.state)
 
     return this
   },
 
-  shouldHistoryKeep(transaction) {
-    return false
-  },
-
-  clean(transaction) {
-    if (transaction.complete && !this.shouldHistoryKeep(transaction)) {
-      this.base = dispatch(this.stores, this.base, transaction)
-      return true
-    }
-    return false
-  },
-
-  transactionWillOpen(transaction) {
-    this.history.append(transaction)
-  },
-
-  transactionWillUpdate(transaction, error, payload, complete) {
-    transaction.active   = !error
-    transaction.complete = complete
-    transaction.error    = error
-    transaction.payload  = payload
-  },
-
-  transactionWillClose() {
-    this.history.prune(this.clean, this)
+  lifecycle(type, initial={}) {
+    return this.plugins.reduce((state, plugin) => {
+      return plugin[type] ? plugin[type](this, state) : state
+    }, initial)
   },
 
-  /**
-   * Resolves an action. As that action signals changes, it will update
-   * a unique transaction. If an error occurs, it will mark it for clean up
-   * and the change will disappear from history.
-   */
   push(action, params, callback) {
-    if (process.env.NODE_ENV !== 'production' && action == null) {
-      throw new TypeError(`Unable to perform: app.push(${ action })\n\n`
-                          + 'This typically happens when an action is accessed from the wrong key of an object, such as:\n\n'
-                          + '• Actions.mispelledAction\n'
-                          + '• import { mispelledAction } from "actions"')
-    }
+    let transaction = new Transaction(tag(action))
 
-    let transaction = Transaction(tag(action))
-    let body = action.apply(null, flatten(params))
+    this.lifecycle(lifecycle.willOpenTransaction, transaction)
 
-    this.transactionWillOpen(transaction)
-
-    return coroutine(body, (error, payload, complete) => {
-      this.transactionWillUpdate(transaction, error, payload, complete)
-
-      if (complete) {
-        this.transactionWillClose(transaction)
-        eventually(callback, this, error, payload)
-      }
-
-      this.rollforward()
-    })
+    return transaction.execute(params, this.rollforward, callback, this)
```

Comment:
Yes.... Eventually I want this to be:

```
transaction.execute(params, 
                    this.transactionDidUpdate, 
                    this.transactionDidFail,
                    this.transactionDidComplete)
```

But I didn't want to totally re-write co-routine at the same time.